Add docs for building on Apple Silicon

geoffrey4444 opened the pull request:
## Proposed changes

Adds instructions for building spectre on Apple Silicon.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Add docs for building on Apple Silicon

MarloMo commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+# Install miniforge. Accept all default choices
+bash Miniforge3-MacOSX-arm64.sh
+# Activate conda base environment for current session
+eval "$(/Users/gwpacworkstation/miniforge3/bin/conda shell.zsh hook)"
```
Would it be better to change to `"$(/Users/<username>/miniforge3/bin/conda shell.zsh hook)"` ? I remember small things like this were confusing for me when I first started.

geoffrey4444 replied:
Good catch! I meant to make such a change but missed it

MarloMo commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+exit # close shell. conda will automatically load when new zsh shells start.
+~~~
+
+Open a new terminal window. Then, install the necessary python packages. (Note:
```
After closing the shell, is it necessary to go back to `apps/src`? With the following steps:
```
export SPECTRE_DEPS_ROOT=$HOME/apps
cd $SPECTRE_DEPS_ROOT/src
```

geoffrey4444 replied:
No, you only need to do that in step 4...the brew and conda commands work from anywhere

nilsvu commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+### 2. Install python dependencies
+
+Spectre depends on python and some python packages. Miniforge is a way to
+install an arm64-native python stack on Apple Silicon Macs. Download
```
Doesn't Homebrew provide a working Python binary for ARM? That would eliminate this step and remove quite a bit of complexity.

geoffrey4444 replied:
They might now but didn't when I was figuring out a working build strategy

kidder replied:
I can brew install numpy and scipy (which of course installed python as well)

geoffrey4444 replied:
Some users (like me) might prefer Miniforge, but I'm glad to know that home-brew also works to install native python. I'll update the docs to show how to install python either way

nilsvu replied:
no need to show multiple ways I think. I thought Homebrew python would make the instructions easier, so I suggested it. Just choose one way.

nilsleiffischer replied:
^

nilsvu commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+Open a new terminal window. Then, install the necessary python packages. (Note:
+jupyter and notebook are recommended, not required.)
+~~~
+conda install numpy scipy matplotlib h5py pyyaml jupyter notebook
```
I recommend to install `jupterlab` instead of `jupyter` and `notebook` these days.

If you can install Python from Homebrew, then replace `conda` with `pip` here. Also move below the Homebrew installation step.

geoffrey4444 replied:
I prefer vscode to jupyterlab, but I understand others might prefer jupyterlab. I'll update the instructions to show both alternatives

nilsvu replied:
I also prefer VSCode. My understanding is that `jupyterlab` is the recommended way to install Jupyter and its dependencies, also when using the VSCode frontend: https://jupyter.org/install

nilsleiffischer replied:
^

geoffrey4444 replied:
Maybe, except to my knowledge apple silicon needs mini forge for native python...maybe that's changed, though

nilsvu commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+### 4. Install remaining dependencies
+
+Here, install the remaining dependencies that cannot be installed
+with homebrew or miniforge.
```
I think that ideally we would merge this whole page into `Installation.md`, since it will surely become out-of-date rather soon. We don't have to repeat explanations on how to clone spectre, how to build charm, how to create build directories, how to run cmake, etc. I restructured the installation page a bit in #3712 to accommodate this better.

Then, the remaining ARM-specific instructions could be added as notes to the installation page, if they are not already taken care of by the corresponding Spack packages. For instance, the `-march=native` issue with libsharp is already handled automatically by this Spack patch: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/libsharp/1.0.0-arm.patch. Therefore, I think the Spack instructions in #3712 might already be sufficient to install the dependency tree on Apple Silicon, though I haven't tested it of course. Could you perhaps give it a try @geoffrey4444?

This page is of course a lot more hands-on tutorial-style than the `Installation.md` page (even with #3712). I suggest that instead of maintaining these step-by-step installation tutorials for a specific system we add enough details to the `Installation.md` page so new people can follow it. We already have Docker, Singularity and Spack instructions that are reasonably system-independent. Manual installation instructions depend so much on the specific system that I don't think we can maintain them. It's each dependency's job to give installation instructions, and we already link to them. What do you think @geoffrey4444?

geoffrey4444 replied:
I thought these instructions would be useful to have around in this level of detail primarily for the Fullerton undergrads, since we have a bunch of M1 Macs. I'd be willing to support it, but if you think a separate page isn't a good idea, I'm fine closing this PR

kidder replied:
spack cannot yet build charm++ for a M1 Mac...

kidder replied:
nor libxsmm...

geoffrey4444 replied:
Have you had any success with the instructions as posted in this PR @kidder ?

kidder replied:
the only thing I was able to install with spack was libsharp; had to install blaze and brigand by hand as well

nilsvu replied:
ok let's add this page then, though I think it would be nice going forward to (1) contribute fixes for ARM builds upstream to Spack and (2) merge these detailed instructions into the main `Installation.md` page.

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+git clone https://github.com/UIUC-PPL/charm
+pushd charm
+git checkout v7.0.0
+./build charm++ multicore-darwin-arm8 --with-production -g3 -j
```
target should be LIBS instead of charm++ to get load-balancing capabilities from charm

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+-D LIBXSMM_ROOT=${SPECTRE_DEPS_ROOT}/libxsmm/ \
+-D BLAZE_ROOT=${SPECTRE_DEPS_ROOT}/blaze/ \
+-D BRIGAND_ROOT=${SPECTRE_DEPS_ROOT}/brigand/ \
+-D LIBSHARP_INCLUDE_DIRS=${SPECTRE_DEPS_ROOT}/libsharp/include \
```
you shouldn't need to specify the include_dirs if you specify root

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+-D LIBSHARP_INCLUDE_DIRS=${SPECTRE_DEPS_ROOT}/libsharp/include \
+-D LIBSHARP_ROOT=${SPECTRE_DEPS_ROOT}/libsharp/ \
+-D Boost_INCLUDE_DIRS=$(brew --prefix boost)/include/ \
+-D Boost_LIBRARIES=$(brew --prefix boost)/lib/ \
```
shouldn't need to specify include_dirs or libraries when specifying root

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,163 @@
+
+~~~
+cmake -D CMAKE_C_COMPILER=clang -D CMAKE_CXX_COMPILER=clang++ -D \
+CMAKE_Fortran_COMPILER=gfortran -D BUILD_PYTHON_BINDINGS=OFF \
```
I was able to build the python bindings

geoffrey4444 replied:
I was as well, so I updated the docs to suggest turning python bindings on

kidder reviewed (changes requested):
I have an apparently working environment on my new laptop, noting what I did differently.  I expect that following the directions as given would also work.  

My requested changes are:
- specify `LIBS` instead of `charm++` when building charm++
- removing the unnecessary definitions from the cmake command for building spectre

geoffrey4444 commented:
Thank you very much for the feedback, everyone! I updated the docs in response to your suggestions and also fixed the doc-check error (which came from auto-formatting to 80-characters, which ended up splitting \subpage and the page name to a separate line, which isn't allowed, it turns out

nilsleiffischer commented:
See comments below (feel free to edit the existing fixup)

nilsvu commented:
See comments: https://github.com/sxs-collaboration/spectre/pull/3714#discussion_r783311245, https://github.com/sxs-collaboration/spectre/pull/3714#discussion_r783312030 (feel free to edit the existing fixup)

nilsvu commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,192 @@
+-D BLAS_ROOT=$(brew --prefix openblas) -D CMAKE_BUILD_TYPE=Debug \
+-D DEBUG_SYMBOLS=ON -D USE_PCH=ON -D SPECTRE_UNIT_TEST_TIMEOUT_FACTOR=5 \
+-D SPECTRE_INPUT_FILE_TEST_TIMEOUT_FACTOR=5 \
+-D SPECTRE_PYTHON_TEST_TIMEOUT_FACTOR=5 \
```
You can replace the three timeout options with the single `SPECTRE_TEST_TIMEOUT_FACTOR` now. It is used as default for all three.

geoffrey4444 commented:
Thank you for the feedback, and my apologies for the delay! I just pushed a fixup incorporating the 3 suggestions @nilsvu mentioned on January 12 and January 15 above.

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,170 @@
+to the x86-64 architecture that SpECTRE usually targets. These instructions
+will result in an Apple Silicon native build of SpECTRE.
+
+\note Floating-point exception trapping and python bindings are not currently
```
remove `and python bindings`

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,170 @@
+Here, install the remaining dependencies that cannot be installed
+with homebrew or miniforge.
+
+Create a directory (such as `$HOME/apps`) to store the build dependencies.
```
Rendering fails from the `$HOME`  on so I cannot review...

geoffrey4444 replied:
This renders for me, but I reworded to avoid the $ inside the backquotes. Does it now render for you?

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -0,0 +1,170 @@
+Name this directory whatever you like, e.g. `$HOME/apps`.
+~~~
+export SPECTRE_DEPS_ROOT=$HOME/apps
+mkdir $SPECTRE_DEPS_ROOT
```
the dollar signs are being rendered as `\f$`

geoffrey4444 replied:
This is fixed by using backquotes instead of tildes to bracket code blocks

geoffrey4444 commented:
Sorry for letting this slide for so long, @kidder ! Could you please take a look when you get a chance? @knelli2 expects that these instructions will be crucial for apple silicon users at the workshop at ICERM that he is leading.

kidder commented on docs/Installation/InstallationOnAppleSilicon.md:
````diff
@@ -9,28 +9,30 @@ on Apple Silicon Macs. Apple Silicon is an arm64 architecture, in contrast
+These instructions, as an example, set this to the `apps` directory in the
+user's home folder.
+```
+export SPECTRE_DEPS_ROOT $HOME/apps
````
need an `=` for export

kidder reviewed (changes requested):
you can squash including the missing `=`
